crates/turborepo-ui/src/output.rs

```diff
@@ -0,0 +1,324 @@
+use std::{
+    borrow::Cow,
+    cell::RefCell,
+    io::{self, Write},
+    sync::{Arc, Mutex},
+};
+
+/// OutputSink represent a sink for outputs that can be written to from multiple
+/// threads through the use of Loggers.
+pub struct OutputSink<W> {
+    writers: Arc<Mutex<SinkWriters<W>>>,
+}
+
+struct SinkWriters<W> {
+    out: W,
+    err: W,
+}
+
+/// OutputClient allows for multiple threads to write to the same OutputSink
+pub struct OutputClient<W> {
+    behavior: OutputClientBehavior,
+    buffer: Option<RefCell<Vec<SinkBytes<'static>>>>,
+    writers: Arc<Mutex<SinkWriters<W>>>,
+}
+
+pub struct OutputWriter<'a, W> {
+    logger: &'a OutputClient<W>,
+    destination: Destination,
+}
+
+/// Enum for controlling the behavior of the client
+#[derive(Debug, Clone, Copy)]
+pub enum OutputClientBehavior {
+    /// Every line sent to the client will get immediately sent to the sink
+    Passthrough,
+    /// Every line sent to the client will get immediately sent to the sink,
+    /// but a buffer will be built up as well and returned when finish is called
+    InMemoryBuffer,
+    // Every line sent to the client will get tracked in the buffer only being
+    // sent to the sink once finish is called.
+    Grouped,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Destination {
+    Stdout,
+    Stderr,
+}
+
+#[derive(Debug, Clone)]
+struct SinkBytes<'a> {
+    buffer: Cow<'a, [u8]>,
+    destination: Destination,
+}
+
+impl<W: Write> OutputSink<W> {
+    /// Produces a new sink with the corresponding out and err writers
+    pub fn new(out: W, err: W) -> Self {
+        Self {
+            writers: Arc::new(Mutex::new(SinkWriters { out, err })),
+        }
+    }
+
+    /// Produces a new client that will send all bytes that it receives to the
+    /// underlying sink. Behavior of how these bytes are sent is controlled
+    /// by the behavior parameter. Note that OutputClient intentionally doesn't
+    /// implement Sync as if you want to write to the same sink
+    /// from multiple threads, then you should create a logger for each thread.
+    pub fn logger(&self, behavior: OutputClientBehavior) -> OutputClient<W> {
+        let buffer = match behavior {
+            OutputClientBehavior::Passthrough => None,
+            OutputClientBehavior::InMemoryBuffer | OutputClientBehavior::Grouped => {
+                Some(RefCell::new(Vec::new()))
+            }
+        };
+        let writers = self.writers.clone();
+        OutputClient {
+            behavior,
+            buffer,
+            writers,
+        }
+    }
+}
+
+impl<W: Write> OutputClient<W> {
```

Comment:
Does it make sense for us to implement https://docs.rs/futures/latest/futures/sink/trait.Sink.html on this? That way we can fairly trivially take a stream (say stdin), map, filter, etc and route the stream directly to the Sink

Reply:
I think in the future if we start using the async IO trait variants (`AsyncWrite` etc) that will definitely be the route we want to go.